[kynhan] iP

kynhan opened the pull request:
## **Duke**

> “Your mind is for having ideas, not holding them.” – David Allen [source](https://dansilvestre.com/productivity-quotes)

Duke frees your mind of having to remember things you need to do. It's
- text-based
- easy to learn :+1:
- ~~FAST~~ *SUPER FAST* to use

All you need to do is
1. Download it from [here](https://github.com/kynhan/ip/releases/tag/A-Jar)
2. Run it from terminal using the command java -jar

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public static void main(String[] args) {
    Duke duke = new Duke();
    duke.run();
}
```

Commit:
Add Gradle support

kynhan committed:
Level-1

kynhan committed:
Level-2

kynhan committed:
Level-3

kynhan committed:
Level-4

kynhan committed:
A-TextUITesting

kynhan committed:
Level-5

kynhan committed:
Level-6

kynhan committed:
Level-7

kynhan committed:
Level-8

kynhan committed:
Merge branch 'branch-Level-8'

kennethk-1201 commented on src/main/java/Event.java:
```diff
@@ -1,12 +1,16 @@
 
-    private String from;
-    private String to;
+    private LocalDateTime fromDate;
+    private LocalDateTime toDate;
+    private final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("d/MM/yyyy H:mm");
```
Nice variable names 😀

kennethk-1201 commented on src/main/java/Task.java:
```diff
@@ -22,7 +22,7 @@ public String toString() {
 
     public void setDone(boolean val) {
-        this.done = val;
+        done = val;
```
Clean code!

CloudHill commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,221 @@
+                deleteTask(input);
+            } else {
+                throw new DukeException("I DON'T UNDERSTAND THAT!");
+            }
```
Your method names seem consistent, but maybe you can have `markTask` and `unmarkTask` instead to differentiate the `show` actions that do not affect your data?

CloudHill commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,87 @@
+                case "T":
+                    Task t = new Task(descrip);
+                    t.setDone(marked);
+                    taskArr.add(t);
```
These two lines look like they're being repeated in the other cases.
Since `Deadline` and `Event` are subclasses of `Task`, maybe you can use polymorphism to factor this out?

CloudHill commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,221 @@
+        }
+
+        // Get to-do name
+        String todo = input.substring(5);
```
It might more sense to name `todo` as `description` or `taskDesc`, since that's the field name in `Task` (maybe also for the `taskName`s in `addDeadline` and `addEvent`)

---

Also, a pretty cool trick with `split` is that you can pass a `limit` as the second parameter, so you can do this to extract just the first word:

```java
String input = "todo trash the trash";
String[] arr = input.split(" ", 2); // only split to 2 elements
arr[0]; // "todo"
arr[1]; // "trash the trash"
```

So you can do something like this if you want to:

```suggestion
        String description = input.split(" ", 2)[1];
```

It's a bit longer, but at least you won't have to hardcode the number of characters.

(I would actually recommend doing the splitting after line 35 so that you can separate the command and arguments immediately)

CloudHill commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,221 @@
+        System.out.println(horizontalLine);
+        System.out.println("HERE ARE YOUR TASKS!");
+        for (int i = 0; i < taskArr.size(); i++) {
+            System.out.println("" + (i+1) + ". " +
+                    "[" + taskArr.get(i).getIcon() + "]" +
+                    "[" + taskArr.get(i).getStatusIcon() + "] " +
+                    taskArr.get(i));
```
You can use `String.format` here if you want (and anywhere else with a lot of concatenation)

Also, since you're using `taskArr.get(i)` a lot, maybe you can store it in a variable.
```suggestion
            Task task = taskArr.get(i);
            System.out.println(String.format(
                    "%d. [%s] [%s] %s",
                    i + 1, task.getIcon(), task.getStatusIcon(), task.toString()));
```

CloudHill reviewed (commented):
Overall, your code looks good! Just a few nits and a couple tricks you can try if you want to.

kynhan committed:
Added more OOP

kynhan committed:
Added Duke package

kynhan committed:
Merge branch 'add-gradle-support' of https://github.com/kynhan/ip

kynhan committed:
Added gradle support

kynhan committed:
Added JUnit tests

kynhan committed:
Corrected mainClassName in build.gradle

EksdeeeX commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,34 @@
+public class Task {
+
+    protected String description;
+    protected boolean done;
```
maybe renaming the variable done to "isDone" would better suit the coding standards?

EksdeeeX commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,51 @@
+
+public class TaskList {
+
+    private ArrayList<Task> taskArr;
```
I like how in the variable name you specified what the ArrayList is containing. It made me realise the areas of my code which I can improve on.

EksdeeeX commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,46 @@
+package duke;
```
Overall good conformation to coding standards, good job.

EksdeeeX commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,121 @@
+        String[] inputs2 = inputs1[1].split(" /to ");
+        String from = inputs2[0];
+        String to = inputs2[1];
+
```
I like the clear description given to the variables

EksdeeeX commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,101 @@
+        System.out.println(horizontalLine);
+    }
+
+}
```
clear method names, good job.

kynhan committed:
Added JavaDocs

kynhan committed:
Made sure code complies with coding standard

kynhan committed:
Started find feature

kynhan committed:
Added find command

kynhan committed:
Started GUI

kynhan committed:
Added GUI

kynhan committed:
fix

kynhan committed:
Add assertions

Add assertions to the Command class to make sure each command
receives the correct number of data arguments.

kynhan committed:
Improve code quality

Split long methods and improve boolean variable names.

kynhan committed:
Merge pull request #2 from kynhan/branch-A-Assertions

Add Assertions

kynhan committed:
Merge master

Merge branch 'master' into branch-A-CodeQuality

kynhan committed:
Merge pull request #3 from kynhan/branch-A-CodeQuality

Improve Code Quality

kynhan committed:
Add fixed duration tasks

kynhan committed:
Add Ui image

kynhan committed:
Update README.md

kynhan committed:
Fix filepath of storage

kynhan committed:
Add user guide

kynhan committed:
Fix user guide

kynhan committed:
Fix user guide typo

kynhan committed:
Remove unneccessary code

kynhan committed:
Fixed build.gradle mainClassName
